app/components/inputs/hooks/useUploadImage.ts

```diff
@@ -0,0 +1,143 @@
+import { ImageListType, ImageType } from "react-images-uploading";
+import { useState } from "react";
+import axios from "axios";
+import { toast } from "react-hot-toast";
+import { PreviewImageWithUrl } from "@/app/types";
+
+interface Props {
+  value: PreviewImageWithUrl[];
+  onChange: (value: PreviewImageWithUrl[]) => void;
+}
+
+export const useUploadImage = ({ value: images, onChange }: Props) => {
+  const [previewImages, setPreviewImages] = useState(
+    images.map((image) => {
+      const { url, ...rest } = image;
+      return { ...rest, dataURL: url as string } as ImageType;
+    })
+  );
+
+  const [isUploading, setIsUploading] = useState(false);
+
+  const uploadImage = async (imageData: string) => {
+    const axiosResponse = await axios.post("/api/images", {
+      imageData,
+    });
+    return axiosResponse.data.url as string;
+  };
+
+  const onImageFileChange = async (
+    imageList: ImageListType,
+    addUpdateIndex: number[] | undefined
+  ) => {
+    console.log(imageList);
+    setPreviewImages(imageList);
+
+    // delete image
+    if (addUpdateIndex === undefined) {
+      const remainingImages = images.filter((image) =>
+        imageList.find((v) => v.dataURL === image.url)
+      );
+      onChange(remainingImages);
+      setPreviewImages(
+        remainingImages.map((image) => {
+          const { url, ...rest } = image;
+          return { ...rest, dataURL: url as string };
+        })
+      );
+      return;
+    }
+
+    setIsUploading(true);
+
+    // replace image
+    if (images.length && addUpdateIndex[0] < images.length) {
```

Comment:
why checking `addUpdateIndex[0]` specifically?

Reply:
Because the numbers in the array are in increasing order. If the first one falls onto an existing index, I can be sure I'm replacing an image, not adding new ones.